src/main/java/org/reactome/addlinks/db/ReferenceObjectCache.java

```diff
@@ -106,183 +106,217 @@ private static synchronized void buildReferenceCaches(String className, Map<Stri
+					allSpecies = new HashSet<>(referringRefTranscripts.size());
+					// Populate a list of species that this ReferenceRNASequence could be cached by.
+					for (GKInstance refTranscript : referringRefTranscripts)
 					{
-						String identifier = (String) referenceObject.getAttributeValue(ReactomeJavaConstants.identifier);
 						try
 						{
-							if (!objectCacheByIdentifier.containsKey(identifier))
-							{
-								objectCacheByIdentifier.put(identifier, Collections.synchronizedList(new ArrayList<GKInstance>(Arrays.asList(referenceObject))) );
-							}
-							else
-							{
-								objectCacheByIdentifier.get(identifier).add(referenceObject);
-							}
+							allSpecies.add(  ((GKInstance)refTranscript.getAttributeValue(ReactomeJavaConstants.species)).getDBID().toString() );
 						}
-						catch (ArrayIndexOutOfBoundsException e)
+						catch (InvalidAttributeException e)
 						{
-							logger.error("ArrayIndexOutOfBounds was caught! Identifier that triggered this was: {}; className: {}; objectCacheByIdentifier has {} items; objectCacheByIdentifier[identifier] has {} items", identifier, className, objectCacheByIdentifier.size(), objectCacheByIdentifier.containsKey(identifier) ? objectCacheByIdentifier.get(identifier).size() : " *THAT IDENTIFIER IS NOT IN THAT MAP!* ");
-							e.printStackTrace();
-							throw new Error(e);
+							logger.error("Could not get the \"{}\" attribute for {} because it not valid for this object.", ReactomeJavaConstants.species, refTranscript);
+						}
+						catch (Exception e)
+						{
+							logger.error("Could not get the \"{}\" attribute for {}. Reason: {}", ReactomeJavaConstants.species, refTranscript, e.getMessage());
 						}
-
 					}
 				}
-				catch (InvalidAttributeException e)
+				catch (Exception e1)
 				{
-					logger.error("Object {} does not have an identifier attribute, error: {}", referenceObject, e.getMessage());
-					e.printStackTrace();
+					logger.error("Could not get ReferenceTranscripts for {} (while trying to determine the species)", referenceObject);
 				}
-				catch (Exception e)
+				
+			}
+		}
+
+		// If the list wasn't initialized yet, it means that we are probably dealing with a ReferenceGeneProduct or a ReferenceDNASequence which should
+		// have its own species.
+		if (allSpecies == null)
+		{
+			allSpecies = new HashSet<>(1);
+			String species;
+			try
+			{
+				species = String.valueOf( ((GKInstance) referenceObject.getAttributeValue(ReactomeJavaConstants.species)).getDBID() );
+				allSpecies.add(species);
+			}
+			catch (InvalidAttributeException e)
+			{
+				logger.error("Could not get the \"{}\" attribute for {} because it not valid for this object.", ReactomeJavaConstants.species, referenceObject);
+			}
+			catch (Exception e)
+			{
+				logger.error("Could not get the \"{}\" attribute for {}. Reason: {}", ReactomeJavaConstants.species, referenceObject, e.getMessage());
+			}
+			
+		}
+		if (cacheBySpecies != null)
```

Comment:
Would this mean the method would not cache the object if the parameter is null?

Reply:
The object will not get added to a species-based cache. This is done with ReferenceMolecules and ReferenceTherapeutics and DatabaseIdentifiers, because they do not have a species attribute, so caching them based on their species would not make sense. So if the species cache object passed in is null, then species-based caching will not happen.

Reply:
Okay, just wanted to check the method shouldn't do something if it receives a null value for that parameter.  Is it worth checking that upfront in the method so you don't have to do anything else if the method receives a null?

Reply:
It might be worth moving the check to the beginning. This code has been refactored many times - at one time I think this code was a part of a larger method that did other stuff even when that specific parameter was null.